##### <h1 style="text-align:center"> Drexel University </h1>
<h2 style = "text-align:center"> College of Computing and Informatics</h2>
<h2 style = "text-align:center">DSCI 631: Applied Machine Learning</h2>
<h3 style = "text-align:center">Assignment 2</h3>
<h4> </h4>
<div style="text-align:center; border-style:solid; padding: 10px">
<div style="font-weight:bold">Due Date: Sunday, February 21, 2021</div>
This assignment counts for 15% of the final grade
</div>

<h3 style="color:red; font-weight:bold; text-decoration: underline">DON'T FORGET TO PUT YOUR TEAM NUMBER AND MEMBERS' NAMES BELOW</h3>
### TEAM NUMBER: 
Group 1 <br>
### TEAM MEMBERS: 
Dustin Luchmee, Laura Quinlan, Toby Sheung <br>

### A. Assignment Overview
This assignment provides the opportunity for you to practice with various skills in feature analysis, feature selection, binary and multi-class classification (imbalance handling) and model tuning.

<h4 style="color:blue; font-weight:bold">The goal of this assignment is to provide a realistic setting for a machine learning task. Therefore instructions will not specify the exact steps to carry out. Instead, it is part of the assignment to identify promising features, models and preprocessing methods and apply them as appropriate.</h4>


### B. What to Hand In
	
Sumbit a completed this Jupyter notebook. 

### C. How to Hand In

Submit your Jupyter notebook file through the course website in the Blackboard Learn system.

### D. When to Hand In

1. Submit your assignment no later than 11:59 pm in the due date.
2. There will be a 10% (absolute value) deduction for each day of lateness, to a maximum of 3 days; assignments will not be accepted beyond that point. Missing work will earn a zero grade.

### E. Written Presentation Requirements (if applicable)
Images must be clear and legible. Assignments will be judged on the basis of visual appearance, grammatical correctness, and quality of writing, as well as their contents. Please make sure that the text of your assignments is well-structured, using paragraphs, full sentences, and other features of well-written presentation.

### F. Academic Honesty

Each student is required to submit the Academic Honesty Form at the beginning of the term to cover all the deliverables (for example: assignments, projects, quizzes). Each piece of work must be original. That means, individual quizzes must be done individually without discussing and collaborating with anybody else. Team assignments must be written and programmed by your own team members. No team should copy any piece of work from other teams. The Drexel University Academic Honesty Rules and Procedures (as stated in the student handbook) will be adhered to strictly.  

### G. Marking Schemes:

Marking assignments will be based on several aspects: presentation, correctness and coding styles. 

For programming questions, 10% of the mark will be judged on the coding style. 

The following is a set of guidelines for the coding style in this course:
1. Write a good comment.
2. Use appropriate indentations to indicate control flows and blocks of code. 
3. When breaking up a long line, break it before an operator, not after. 

### H. Answer the following questions: 
**Your answer should be combined with code and brief text answer.** Please ensure that your Jupyter notebook does not have too much spurious output. If you like, you can share your notebook in progress with me on Kaggle: leiwangv (lw474@drexel.edu)


### Data for Binary Classification in this assignment:
* URL: https://www.kaggle.com/lodetomasi1995/income-classification
* On Kaggel Notebook, you can add the data set by searching the above URL
* Column “income” is the target label to classify.



In [100]:
# On Kaggle Notebook, after adding the data, you can import the data as Pandas DataFrame as follows
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
df = pd.read_csv("data/income_evaluation.csv")

#### Question 1-1:
Examine the features of the data. Identify which feature is continuous and which feature is categorical. Make some analyses and statistics, then use the results to discuss your selected predictors. You may also derive new features.

In [1]:
df.head()

In [2]:
df.info()

In [103]:
df.columns

Index(['age', ' workclass', ' fnlwgt', ' education', ' education-num',
       ' marital-status', ' occupation', ' relationship', ' race', ' sex',
       ' capital-gain', ' capital-loss', ' hours-per-week', ' native-country',
       ' income'],
      dtype='object')

**Continuous:** age, fnlwgt, education-num, capital-gain, capital-loss, hours-per-week <br>
**Categorical:** workclass, education, martial-status, occupation, relationship, race, sex, native-country, income (target) 

In [3]:
%matplotlib inline

#Plotting the numerical variables
df.hist(figsize=(15,15))

**Observations:** <br>
The scaling of the continuous variables need to be scaled to similar numbers. There are some right skewness with features such as age, fnlwgt and capital-gain. Perhaps a log transformation will be used to create a normal distribution of the data.

In [4]:
#Plotting the categorical variables
df[' workclass'].value_counts().plot(kind = 'bar', figsize=(6,4))

In [5]:
df[' workclass'].value_counts()

**Observations:** <br>
Most of the workclass instances are private. The feature could possibly be omitted...

In [6]:
df[' education'].value_counts().plot(kind = 'bar', figsize=(7,4))

**Observation:** <br>
Both education and education-num seem to be similar features about the same concept. The only difference is that one is numerical and another is categorical and more specific about the education level.

In [7]:
df[' marital-status'].value_counts().plot(kind = 'bar', figsize=(7,4))

In [8]:
df[' occupation'].value_counts().plot(kind = 'bar', figsize=(7,4))

In [9]:
df[' relationship'].value_counts().plot(kind = 'bar', figsize=(7,4))

In [10]:
df[' race'].value_counts().plot(kind = 'bar', figsize=(7,4))

In [112]:
df[' race'].value_counts()

 White                 27816
 Black                  3124
 Asian-Pac-Islander     1039
 Amer-Indian-Eskimo      311
 Other                   271
Name:  race, dtype: int64

In [11]:
df[' sex'].value_counts().plot(kind = 'bar', figsize=(7,4))

In [114]:
df[' sex'].value_counts()

 Male      21790
 Female    10771
Name:  sex, dtype: int64

In [12]:
#Barplot of sex and income level
sns.countplot(data = df, x = ' income', hue = ' sex')

In [13]:
#Examination of how sex affects the other numerical features
sns.pairplot(data = df, hue = ' sex')

**Observation:** <br>
From the above plot analysis, it is highly likely that sex would play a large part in whether or not someone makes more than 50K. Although there are more male data instances, there are 10,000 instances that are female.  

In [14]:
df[' native-country'].value_counts().plot(kind = 'bar', figsize=(7,4))

In [15]:
df[' native-country'].value_counts()

**Observations:** <br>
When examining the categorical features, the most unlikely useful variable to help create a good model is the native-country feature. Most of instances are from the United-States. Another feature that may not be too helpful is the race variable. Much of the data are those that are white. 

In [119]:
#Examination of missing values
df.isin([' ?']).sum()

age                   0
 workclass         1836
 fnlwgt               0
 education            0
 education-num        0
 marital-status       0
 occupation        1843
 relationship         0
 race                 0
 sex                  0
 capital-gain         0
 capital-loss         0
 hours-per-week       0
 native-country     583
 income               0
dtype: int64

**Observations:** <br>
When using the info() for this dataset, I was surprised to see that there were no missing values. But a closer examination of the values in the occupation barplot shows that ? is used as a placeholder. Both workclass and occupation have a close number of missing values and the variables are similar to each other. The next step is to examine how many missing value indexes within workclass and occupation are the same instance. 

In [120]:
#Captures index of workclass index with missing values
work_array = df[' workclass'].isin([' ?'])
work_miss = np.where(work_array)[0]

In [121]:
#Captures index of occupation index with missing values
occup_array = df[' occupation'].isin([' ?'])
occup_miss = np.where(occup_array)[0]

In [122]:
#Quick examination of the missing index of workclass and occupation
print(work_miss)
print(occup_miss)

[   27    61    69 ... 32539 32541 32542]
[   27    61    69 ... 32539 32541 32542]


In [123]:
#Print number of matching missing indexes for workclass and occupation
len(set(work_miss) & set(occup_miss))

1836

**Observations:** <br>
The above analysis shows that the all of the workclass missing instances are the same in the occupation feature. There are 1836 missing instances that have both missing workclass and occupation information. Perhaps when hot encoding, creating a new category for the missing value would be a good choice. 

In [124]:
#Copying the dataset to use the dummy encoder to test out feature selection model
copy_df = df.copy()
copy_df = copy_df.drop([' native-country'], axis = 1)
copy_df_dum = pd.get_dummies(data = copy_df)
copy_df_dum.head()

age   fnlwgt   education-num   capital-gain   capital-loss  \
0   39    77516              13           2174              0   
1   50    83311              13              0              0   
2   38   215646               9              0              0   
3   53   234721               7              0              0   
4   28   338409              13              0              0   

    hours-per-week   workclass_ ?   workclass_ Federal-gov  \
0               40              0                        0   
1               13              0                        0   
2               40              0                        0   
3               40              0                        0   
4               40              0                        0   

    workclass_ Local-gov   workclass_ Never-worked  ...   relationship_ Wife  \
0                      0                         0  ...                    0   
1                      0                         0  ...                    0   
2                      0                         0  ...                    0   
3                      0                         0  ...                    0   
4                      0                         0  ...                    1   

    race_ Amer-Indian-Eskimo   race_ Asian-Pac-Islander   race_ Black  \
0                          0                          0             0   
1                          0                          0             0   
2                          0                          0             0   
3                          0                          0             1   
4                          0                          0             1   

    race_ Other   race_ White   sex_ Female   sex_ Male   income_ <=50K  \
0             0             1             0           1               1   
1             0             1             0           1               1   
2             0             1             0           1               1   
3             0             0             0           1               1   
4             0             0             1           0               1   

    income_ >50K  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 68 columns]

In [16]:
copy_df_dum.columns

In [126]:
corr = copy_df_dum.corr()
corr[' income_ >50K'].sort_values(ascending = False).head(10)
#Wonder if I can average the same categorical variables

 income_ >50K                          1.000000
 marital-status_ Married-civ-spouse    0.444696
 relationship_ Husband                 0.401035
 education-num                         0.335154
age                                    0.234037
 hours-per-week                        0.229689
 capital-gain                          0.223329
 sex_ Male                             0.215980
 occupation_ Exec-managerial           0.214861
 occupation_ Prof-specialty            0.185866
Name:  income_ >50K, dtype: float64

In [127]:
corr[' income_ <=50K'].sort_values(ascending = False).head(10)

 income_ <=50K                    1.000000
 marital-status_ Never-married    0.318440
 relationship_ Own-child          0.228532
 sex_ Female                      0.215980
 relationship_ Not-in-family      0.188497
 occupation_ Other-service        0.156348
 relationship_ Unmarried          0.142857
 education_ HS-grad               0.131189
 marital-status_ Divorced         0.126995
 occupation_ Adm-clerical         0.089986
Name:  income_ <=50K, dtype: float64

In [128]:
marital_data = copy_df_dum.loc[:,['age', ' fnlwgt', ' education-num', ' capital-gain', ' capital-loss',
       ' hours-per-week', ' marital-status_ Married-civ-spouse', ' income_ >50K']]
marital_data

age   fnlwgt   education-num   capital-gain   capital-loss  \
0       39    77516              13           2174              0   
1       50    83311              13              0              0   
2       38   215646               9              0              0   
3       53   234721               7              0              0   
4       28   338409              13              0              0   
...    ...      ...             ...            ...            ...   
32556   27   257302              12              0              0   
32557   40   154374               9              0              0   
32558   58   151910               9              0              0   
32559   22   201490               9              0              0   
32560   52   287927               9          15024              0   

        hours-per-week   marital-status_ Married-civ-spouse   income_ >50K  
0                   40                                    0              0  
1                   13                                    1              0  
2                   40                                    0              0  
3                   40                                    1              0  
4                   40                                    1              0  
...                ...                                  ...            ...  
32556               38                                    1              0  
32557               40                                    1              1  
32558               40                                    0              0  
32559               20                                    0              0  
32560               40                                    1              1  

[32561 rows x 8 columns]

In [17]:
marital_corr = marital_data.corr()
plt.figure(figsize=(7,7))
sns.heatmap(marital_corr, cmap = 'YlGnBu')

NameError: name 'marital_data' is not defined

**Observations:** <br>
From the above analysis, some important features for the prediction model is sex, martial_status(Married-civ-spouse) and education-num.  

#### Question 1-2:
Check if the target label is balanced or not, what is your strategy if imbalanced? apply your strategy.

In [130]:
df[' income'].value_counts()

 <=50K    24720
 >50K      7841
Name:  income, dtype: int64

In [18]:
df[' income'].value_counts().plot(kind = 'bar')

**Observations:** <br>
The target label is imbalanced and the strategy that will be used on this is the SMOTE technique. SMOTE allows the creation of new instances that are similar to the original data. Unlike upsampling, there will not be repeated instances. But in order to use this technique, the data will have to be split, cleaned and hot encoded before the use of SMOTE to prevent data leakage. This strategy will be applied within the pipeline in the next question.

#### Question 1-3: 
Split the data into train and test data, then build Pipeline, apply transformer as needed, train the following 4 algorithms (Logistic Regression, Support Vector Machines, Naïve Bayes and K-Nearest Neighbors) with hyperparameter tuning, to classify the "income". Compare the performance of different algorithms on Test set. Discuss your results and findings.

For Naïve Bayes, which algortihm you adopted here, why you choose this one?

In [132]:
from sklearn.model_selection import train_test_split

train_set, test_set = train_test_split(df, test_size = 0.2, random_state = 42)

#Set labels
train_labels = train_set[" income"] 
test_labels = test_set[" income"]

#Drop target from training and testing set
true_train_set = train_set.drop(" income", axis = 1)
true_test_set = test_set.drop(" income", axis = 1)

In [133]:
true_train_set.head()

age          workclass   fnlwgt      education   education-num  \
5514    33          Local-gov   198183      Bachelors              13   
19777   36            Private    86459      Assoc-voc              11   
10781   58   Self-emp-not-inc   203039            9th               5   
32240   21            Private   180190      Assoc-voc              11   
9876    27            Private   279872   Some-college              10   

            marital-status        occupation    relationship    race      sex  \
5514         Never-married    Prof-specialty   Not-in-family   White   Female   
19777   Married-civ-spouse   Exec-managerial         Husband   White     Male   
10781            Separated      Craft-repair   Not-in-family   White     Male   
32240   Married-civ-spouse   Farming-fishing         Husband   White     Male   
9876              Divorced     Other-service   Not-in-family   White     Male   

        capital-gain   capital-loss   hours-per-week  native-country  
5514               0              0               50   United-States  
19777              0           1887               50   United-States  
10781              0              0               40   United-States  
32240              0              0               46   United-States  
9876               0              0               40   United-States

In [134]:
true_train_set.columns

Index(['age', ' workclass', ' fnlwgt', ' education', ' education-num',
       ' marital-status', ' occupation', ' relationship', ' race', ' sex',
       ' capital-gain', ' capital-loss', ' hours-per-week', ' native-country'],
      dtype='object')

In [135]:
#Since native-country does not tell much information, it is dropped

true_train_set = true_train_set.drop([" native-country", " workclass", " race", " fnlwgt", " capital-loss"], axis = 1)
true_test_set = true_test_set.drop([" native-country", " workclass", " race", " fnlwgt", " capital-loss"], axis = 1)

In [136]:
true_train_set.head()

age      education   education-num       marital-status  \
5514    33      Bachelors              13        Never-married   
19777   36      Assoc-voc              11   Married-civ-spouse   
10781   58            9th               5            Separated   
32240   21      Assoc-voc              11   Married-civ-spouse   
9876    27   Some-college              10             Divorced   

             occupation    relationship      sex   capital-gain  \
5514     Prof-specialty   Not-in-family   Female              0   
19777   Exec-managerial         Husband     Male              0   
10781      Craft-repair   Not-in-family     Male              0   
32240   Farming-fishing         Husband     Male              0   
9876      Other-service   Not-in-family     Male              0   

        hours-per-week  
5514                50  
19777               50  
10781               40  
32240               46  
9876                40

In [137]:
#Separate categorical and numerical variables for pipeline

cate_attr = [" marital-status", " education", " relationship", " occupation", " sex"]
num_attr = ["age", " education-num", " hours-per-week", " capital-gain"]

In [138]:
#Combining categorical and numerical pipeline

from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder

combine_pipeline = ColumnTransformer([
    ("num", StandardScaler(), num_attr),
    ("cat", OneHotEncoder(handle_unknown = "ignore"), cate_attr),
    ])

In [139]:
#Fit and Transform Train
trans_train = combine_pipeline.fit_transform(true_train_set)


#Transform Test
trans_test = combine_pipeline.transform(true_test_set)

In [140]:
#Create SMOTE algorithm

from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state = 42)

#resample training set
x_smote_train, x_smote_labels = smote.fit_resample(trans_train, train_labels)

### Logistic Regression Model

#### Without specific hyperparameters  

In [141]:
from sklearn.linear_model import LogisticRegression

#Logistic Regression algorithm
log_reg = LogisticRegression(random_state = 42, solver='lbfgs', max_iter = 400)

In [142]:
log_reg.fit(x_smote_train, x_smote_labels)

LogisticRegression(max_iter=400, random_state=42)

In [143]:
#Logisitic Regression Model Score
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix, classification_report

#Predict labels
pred_log_reg = log_reg.predict(trans_test)

#accuracy score
log_acc = accuracy_score(pred_log_reg, test_labels)
#precision score
log_prec = precision_score(pred_log_reg, test_labels, average = 'binary', pos_label = " >50K")
#recall_score
log_recall = recall_score(pred_log_reg, test_labels, average = 'binary', pos_label = " >50K")

print(log_acc, log_prec, log_recall)
print(classification_report(test_labels, pred_log_reg))

0.8053124520190389 0.8453214513049013 0.5643858903527412
              precision    recall  f1-score   support

       <=50K       0.94      0.79      0.86      4942
        >50K       0.56      0.85      0.68      1571

    accuracy                           0.81      6513
   macro avg       0.75      0.82      0.77      6513
weighted avg       0.85      0.81      0.82      6513



#### Automatic hyperparameter search

In [144]:
log_reg = LogisticRegression(random_state = 42)

#Hyperparameter choices
param_grid = [
    {"penalty" : ["l1", "l2", "elasticnet", "none"],
    "solver" : ["newton-cg", "lbfgs", "liblinear", "sag", "saga"],
     "max_iter" : [300, 400, 500, 1000]
    }
]

In [145]:
#Use of GridSearchCV
from sklearn.model_selection import GridSearchCV

hyper_log = GridSearchCV(log_reg, param_grid = param_grid, cv = 3, n_jobs = -1)

In [19]:
find_log = hyper_log.fit(x_smote_train, x_smote_labels)

In [ ]:
#Find best hyperparameters to use with log regression

find_log.best_estimator_

In [ ]:
#Examine accuracy for training set and training labels
find_log.score(x_smote_train, x_smote_labels)

In [ ]:
#Use on testing

log_reg2 = LogisticRegression(random_state = 42, solver = "liblinear", penalty = "l1", max_iter = 300)
log_reg2.fit(x_smote_train, x_smote_labels)

#Predict labels
pred_log_reg = log_reg2.predict(trans_test)

#accuracy score
hyper_log_acc = accuracy_score(pred_log_reg, test_labels)
#precision score
hyper_log_prec = precision_score(pred_log_reg, test_labels, average = 'binary', pos_label = " >50K")
#recall_score
hyper_log_recall = recall_score(pred_log_reg, test_labels, average = 'binary', pos_label = " >50K")

print(hyper_log_acc, hyper_log_prec, hyper_log_recall)
print(classification_report(test_labels, pred_log_reg))

**Observations:** <br>
Even with GridSearchCV and hyperparamter tuning, the results of the accuracy, precision and recall is the same as the first logistic regression model without any specific reasoning with its hyperparameters. Both the accuracy and precision score is not bad, only the recall is slightly lower. 

### Support Vector Machine Model 

#### Without specific hyperparameters

In [148]:
from sklearn.svm import LinearSVC
from sklearn.svm import SVC

#SVM Linear Algorithm
svm = LinearSVC(C = 1, loss = "hinge")
svm.fit(x_smote_train, x_smote_labels)

/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:985: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


LinearSVC(C=1, loss='hinge')

In [149]:
pred_svm = svm.predict(trans_test)

#accuracy score
svm_acc = accuracy_score(pred_svm, test_labels)
#precision score
svm_prec = precision_score(pred_svm, test_labels, average = 'binary', pos_label = " >50K")
#recall_score
svm_recall = recall_score(pred_svm, test_labels, average = 'binary', pos_label = " >50K")

print(svm_acc, svm_prec, svm_recall)

0.7910333179794258 0.852959898154042 0.5425101214574899


#### Automatic hyperparameter search

In [150]:
svm = LinearSVC()

#Hyperparameter choices
param_grid = [
    {"penalty" : ["l1", "l2"],
    "loss" : ["hinge", "squared_hinge"],
     "C" : [1.0, 2.0, 3.0, 0.5]
    }
]

In [151]:
#Create GridSearch for SVM
hyper_lsvm = GridSearchCV(svm, param_grid = param_grid, cv = 3, n_jobs = -1) 

In [152]:
find_l_svm = hyper_lsvm.fit(x_smote_train, x_smote_labels)

exception calling callback for <Future at 0x7fa8846fe6a0 state=finished raised TerminatedWorkerError>
Traceback (most recent call last):
  File "/opt/anaconda3/lib/python3.8/site-packages/joblib/externals/loky/_base.py", line 625, in _invoke_callbacks
    callback(self)
  File "/opt/anaconda3/lib/python3.8/site-packages/joblib/parallel.py", line 347, in __call__
    self.parallel.dispatch_next()
  File "/opt/anaconda3/lib/python3.8/site-packages/joblib/parallel.py", line 780, in dispatch_next
    if not self.dispatch_one_batch(self._original_iterator):
  File "/opt/anaconda3/lib/python3.8/site-packages/joblib/parallel.py", line 847, in dispatch_one_batch
    self._dispatch(tasks)
  File "/opt/anaconda3/lib/python3.8/site-packages/joblib/parallel.py", line 765, in _dispatch
    job = self._backend.apply_async(batch, callback=cb)
  File "/opt/anaconda3/lib/python3.8/site-packages/joblib/_parallel_backends.py", line 531, in apply_async
    future = self._workers.submit(SafeFunction(func))

TerminatedWorkerError: A worker process managed by the executor was unexpectedly terminated. This could be caused by a segmentation fault while calling the function or by an excessive memory usage causing the Operating System to kill the worker.

The exit codes of the workers are {SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6), SIGABRT(-6)}

In [ ]:
find_l_svm.best_estimator_

In [153]:
#Training score
find_l_svm.score(x_smote_train, x_smote_labels)

0.82146829810901

In [155]:
#Use on testing

svm_2 = LinearSVC(C = 0.5)
svm_2.fit(x_smote_train, x_smote_labels)

#Predict labels
pred_svm = svm_2.predict(trans_test)

#accuracy score
hyper_svm_acc = accuracy_score(pred_svm, test_labels)
#precision score
hyper_svm_prec = precision_score(pred_svm, test_labels, average = 'binary', pos_label = " >50K")
#recall_score
hyper_svm_recall = recall_score(pred_svm, test_labels, average = 'binary', pos_label = " >50K")

print(hyper_svm_acc, hyper_svm_prec, hyper_svm_recall)
print(classification_report(test_labels, pred_svm))

0.7994779671426377 0.8491406747294716 0.5551394090719933
              precision    recall  f1-score   support

       <=50K       0.94      0.78      0.86      4942
        >50K       0.56      0.85      0.67      1571

    accuracy                           0.80      6513
   macro avg       0.75      0.82      0.76      6513
weighted avg       0.85      0.80      0.81      6513



### Poly Support Vector Machine Model

#### #1

In [156]:
#Poly SVM
p_svm = SVC(kernel = "poly", degree = 2, coef0 = 1, C = 5)
p_svm.fit(x_smote_train, x_smote_labels)

SVC(C=5, coef0=1, degree=2, kernel='poly')

In [157]:
pred_p_svm = p_svm.predict(trans_test)

#accuracy score
p_svm_acc = accuracy_score(pred_p_svm, test_labels)
#precision score
p_svm_prec = precision_score(pred_p_svm, test_labels, average = 'binary', pos_label = " >50K")
#recall_score
p_svm_recall = recall_score(pred_p_svm, test_labels, average = 'binary', pos_label = " >50K")

print(p_svm_acc, p_svm_prec, p_svm_recall)

0.8062336864732075 0.8478676002546149 0.5656050955414013


#### #2

In [158]:
#Poly SVM
p_svm = SVC(kernel = "poly", degree = 3, coef0 = 2, C = 5)
p_svm.fit(x_smote_train, x_smote_labels)

SVC(C=5, coef0=2, kernel='poly')

In [159]:
pred_p_svm = p_svm.predict(trans_test)

#accuracy score
p_svm_acc2 = accuracy_score(pred_p_svm, test_labels)
#precision score
p_svm_prec2 = precision_score(pred_p_svm, test_labels, average = 'binary', pos_label = " >50K")
#recall_score
p_svm_recall2 = recall_score(pred_p_svm, test_labels, average = 'binary', pos_label = " >50K")

print(p_svm_acc2, p_svm_prec2, p_svm_recall2)

0.8080761553815446 0.8427753023551878 0.568972926514826


#### Gaussian RBF Kernel

In [160]:
#Gaussian RBF Kernel
g_svm = SVC(kernel = "rbf", gamma = 5, C = 0.001)
g_svm.fit(x_smote_train, x_smote_labels)

SVC(C=0.001, gamma=5)

In [161]:
g_pred_svm = g_svm.predict(trans_test)

#accuracy score
g_svm_acc = accuracy_score(g_pred_svm, test_labels)
#precision score
g_svm_prec = precision_score(g_pred_svm, test_labels, average = 'binary', pos_label = " >50K")
#recall_score
g_svm_recall = recall_score(g_pred_svm, test_labels, average = 'binary', pos_label = " >50K")

print(g_svm_acc, g_svm_prec, g_svm_recall)

0.7735298633502227 0.12348822406110757 0.6643835616438356


### Naive Bayes Model

#### #1

In [162]:
from sklearn.naive_bayes import BernoulliNB

bnb = BernoulliNB()
bnb.fit(x_smote_train, x_smote_labels)

BernoulliNB()

In [163]:
pred_bnb = bnb.predict(trans_test)

#accuracy score
bnb_acc = accuracy_score(pred_bnb, test_labels)
#precision score
bnb_prec = precision_score(pred_bnb, test_labels, average = 'binary', pos_label = " >50K")
#recall_score
bnb_recall = recall_score(pred_bnb, test_labels, average = 'binary', pos_label = " >50K")

print(bnb_acc, bnb_prec, bnb_recall)

0.7454322124980808 0.8357733927434755 0.48396608920014744


#### #2

In [164]:
bnb = BernoulliNB(alpha = 2.0)
bnb.fit(x_smote_train, x_smote_labels)

BernoulliNB(alpha=2.0)

In [165]:
pred_bnb = bnb.predict(trans_test)

#accuracy score
bnb_acc2 = accuracy_score(pred_bnb, test_labels)
#precision score
bnb_prec2 = precision_score(pred_bnb, test_labels, average = 'binary', pos_label = " >50K")
#recall_score
bnb_recall2 = recall_score(pred_bnb, test_labels, average = 'binary', pos_label = " >50K")

print(bnb_acc2, bnb_prec2, bnb_recall2)

0.7454322124980808 0.8357733927434755 0.48396608920014744


**Observations:** <br>
The Bernoulli Naive Bayes model was chosen due to its usefulness in binary algorithms. The Bernoulli distribution assumes that there are only two values (0 or 1). It expects binary feature vectors and our dataset uses the one-hot encoding to make the features either 0 or 1. 

### K-Nearest Neighbor Model

#### #1

In [166]:
from sklearn import neighbors

knn = neighbors.KNeighborsClassifier(n_neighbors = 3)
knn.fit(x_smote_train, x_smote_labels)

KNeighborsClassifier(n_neighbors=3)

In [167]:
pred_knn = knn.predict(trans_test)

#accuracy score
knn_acc = accuracy_score(pred_knn, test_labels)
#precision score
knn_prec = precision_score(pred_knn, test_labels, average = 'binary', pos_label = " >50K")
#recall_score
knn_recall = recall_score(pred_knn, test_labels, average = 'binary', pos_label = " >50K")

print(knn_acc, knn_prec, knn_recall)

0.7921080915092891 0.7097390197326544 0.553899652260308


#### #2

In [168]:
knn = neighbors.KNeighborsClassifier(n_neighbors = 2)
knn.fit(x_smote_train, x_smote_labels)

KNeighborsClassifier(n_neighbors=2)

In [169]:
pred_knn = knn.predict(trans_test)

#accuracy score
knn_acc2 = accuracy_score(pred_knn, test_labels)
#precision score
knn_prec2 = precision_score(pred_knn, test_labels, average = 'binary', pos_label = " >50K")
#recall_score
knn_recall2 = recall_score(pred_knn, test_labels, average = 'binary', pos_label = " >50K")

print(knn_acc2, knn_prec2, knn_recall2)

0.8105327805926609 0.5334182049649905 0.6258401792382375


### Overall Metrics

In [170]:
#Creation of metrics dataframe
import pandas as pd

model_metrics = pd.DataFrame([["Logistic Regression", log_acc, log_prec, log_recall], 
                              ["SVM", hyper_svm_acc, hyper_svm_prec, hyper_svm_recall],
                              ["SVM Poly", p_svm_acc2, p_svm_prec2, p_svm_recall2],
                              ["SVM Gaussian", g_svm_acc, g_svm_prec, g_svm_recall],
                              ["Naive Bayes", bnb_acc, bnb_prec, bnb_recall],
                              ["K-Nearest Neighbor", knn_acc2, knn_prec2, knn_recall2]],
                             columns = ["Model", "Accuracy", "Precision", "Recall"])
model_metrics

Model  Accuracy  Precision    Recall
0  Logistic Regression  0.805312   0.845321  0.564386
1                  SVM  0.799478   0.849141  0.555139
2             SVM Poly  0.808076   0.842775  0.568973
3         SVM Gaussian  0.773530   0.123488  0.664384
4          Naive Bayes  0.745432   0.835773  0.483966
5   K-Nearest Neighbor  0.810533   0.533418  0.625840

**Observations:** <br>
K-Nearest Neighbor has the highest accuracy among the model but lower precision. The logistic regression and the poly SVM also did a good job in identifing the target classes and having the highest scores in precision. All of the models have a lower score in recall and have a harder time in the sensitivity of identifying true positives. In this dataset, I believe that having a higher score in accuracy and precision is fine with solving the task of classifing whether or not a person makes more than 50K. 

The lowest model is the SVM Gaussian, especially with such a low precision score. The better support vector machine model is the poly svm or linear svm in this case. 

#### Question 1-4:
Identify features that are important for your best model. Which features are most influential? Which features could be removed without decrease in performance? Does removing irrelevant features make your model better?

**Observations:** <br>
Most influential features: martial-status, education-num, relationship  <br>
Features that could be removed without decrease in performance were fnlwgt, capital-loss, race, native-country <br>
Features removed that increased performance: workclass <br>

The other features such as sex, age, capital-gain, hours-per-week, education, occupation helped a little bit with the model's numbers. When several irrelevant features are removed, the models are run a little faster. Only when removing workclass did the performance of the model increase for the logisitic regression model.

### Data for Multiple-class Classification in this assignment:
* URL: https://www.kaggle.com/uciml/human-activity-recognition-with-smartphones
* On Kaggel Notebook, you can add the data set by searching the above URL
* Column “Activity” is the target label to classify.

In [171]:
# On Kaggle Notebook, after adding the data, you can import the data as Pandas DataFrame as follows
import pandas as pd
df_train = pd.read_csv("data/train.csv")
df_test = pd.read_csv("data/test.csv")

In [172]:
df_train.shape

(7352, 563)

#### Question 2-1:
Create plots to visualize the distribution of the label "Activity". If the distribution of each class is balanced?
If imbalanced, what strategies would you adopt and why? Apply your strategy on the dataset

In [173]:
df_train["Activity"].value_counts()

LAYING                1407
STANDING              1374
SITTING               1286
WALKING               1226
WALKING_UPSTAIRS      1073
WALKING_DOWNSTAIRS     986
Name: Activity, dtype: int64

In [20]:
df_train["Activity"].value_counts().plot(kind = "bar")

**Observations:** <br>
The distribution of the target label is not too imbalanced. But the category "Walking_Downstairs" has 400 less instances than "Laying" the highest instance. Perhaps using a imbalancing strategy would be good to allow all of the categories have an equal footing in the model.

In [175]:
#Examine that only the target variable is not numerical
df_train.select_dtypes("object").columns

Index(['Activity'], dtype='object')

In [176]:
#Train Set
x_train = df_train.drop("Activity", axis = 1)

#Train labels
y_train = df_train["Activity"]

#Test Set
x_test = df_test.drop("Activity", axis = 1)

#Test labels
y_test = df_test["Activity"]

Numerical Values for Target Label

Laying = 0 <br>
Standing = 1 <br>
Sitting = 2 <br>
Walking = 3 <br>
Walking_Upstairs = 4 <br>
Walking_Downstairs = 5 <br>

In [177]:
#Replace categorical target to numerical

cate_to_num = {"LAYING": 0, "STANDING": 1, "SITTING": 2, "WALKING": 3, "WALKING_UPSTAIRS": 4, "WALKING_DOWNSTAIRS": 5}
#Replace train labels
num_y_train = y_train.replace(cate_to_num)

#Replace test labels
num_y_test = y_test.replace(cate_to_num)

In [178]:
#Upsampling data

from imblearn.over_sampling import RandomOverSampler
up = RandomOverSampler()
x_train_up, y_train_up = up.fit_sample(x_train, num_y_train)
print(x_train.shape)
print(x_train_up.shape)
print(y_train_up.shape)

(7352, 562)
(8442, 562)
(8442,)


**Observations:** <br>
The imbalancing strategy used is the Upsampling technique. Due to the overall target labels instances being so low, I chose not to downsample. SMOTE was not used because its use of the euclidean distance to create new instances. The target labels are not ordinal in any way that would allow the use of distances to create useful instances. Upsampling was used because the number of instances created would not be so high and the affect on the model with repeated instances would perhaps be minimal.   

#### Question 2-2:
Evaluate the performance of multi-class classification with 3 self-selected classification algorithm with hyperparameter tuning. 

Of the algorithms you selected, can each of them capable of handling multiple classes natively? Then what strategy/strategies can be applied to perform multiclass classification for those algorithms in order to perform multi-class classification? 

Compare the performance of different algorithms on Test set. 

### Decision Tree Model

#### #1

In [179]:
from sklearn.tree import DecisionTreeClassifier

d_tree = DecisionTreeClassifier(max_depth = 2)
d_tree_model = d_tree.fit(x_train_up, y_train_up)
d_tree_pred = d_tree_model.predict(x_test)

In [180]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(num_y_test, d_tree_pred)
cm

array([[537,   0,   0,   0,   0,   0],
       [  0, 532,   0,   0,   0,   0],
       [  0, 491,   0,   0,   0,   0],
       [  0,   0,   0, 496,   0,   0],
       [  0,   0,   0, 471,   0,   0],
       [  0,   0,   0, 420,   0,   0]])

#### #2

In [181]:
d_tree = DecisionTreeClassifier(max_depth = 3, criterion = "entropy")
d_tree_model = d_tree.fit(x_train_up, y_train_up)
d_tree_pred = d_tree_model.predict(x_test)

In [182]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(num_y_test, d_tree_pred)
cm

array([[537,   0,   0,   0,   0,   0],
       [  0, 425, 107,   0,   0,   0],
       [  0,  91, 400,   0,   0,   0],
       [  0,   0,   0, 437,  48,  11],
       [  0,   0,   0, 129, 336,   6],
       [  0,   0,   0, 107,  52, 261]])

**Observation:** <br>
The first decision tree did not do well in identifying the 2 (Sitting), 4 (Walking_Upstairs) or 5 (Walking_Downstairs) class. The second tree used the entropy criterion and had better classification of each class.  

### SVM

#### #1

In [183]:
from sklearn.svm import SVC

svm = SVC(decision_function_shape='ovo')
svm_model = svm.fit(x_train_up, y_train_up)
svm_pred = svm_model.predict(x_test)

In [184]:
cm = confusion_matrix(num_y_test, svm_pred)
cm

array([[537,   0,   0,   0,   0,   0],
       [  0, 485,  47,   0,   0,   0],
       [  0,  59, 430,   0,   2,   0],
       [  0,   0,   0, 494,   0,   2],
       [  0,   0,   0,  19, 450,   2],
       [  0,   0,   0,  27,  26, 367]])

#### #2

In [185]:
svm = SVC(decision_function_shape='ovo', kernel = "poly", C = 2.0)
svm_model = svm.fit(x_train_up, y_train_up)
svm_pred = svm_model.predict(x_test)

In [186]:
cm = confusion_matrix(num_y_test, svm_pred)
cm

array([[537,   0,   0,   0,   0,   0],
       [  0, 499,  33,   0,   0,   0],
       [  0,  55, 435,   0,   1,   0],
       [  0,   0,   0, 490,   0,   6],
       [  0,   0,   0,  24, 443,   4],
       [  0,   0,   0,  32,  24, 364]])

**Observations:** <br>
Even with the hyperparameter changes, the SVM model has the best score among the other classifiers. The linear svm vs the poly svm has no considerable differences.

### K-Nearest Neighbors

#### #1

In [187]:
from sklearn.neighbors import KNeighborsClassifier 

knn = KNeighborsClassifier(n_neighbors = 7)
knn_model = knn.fit(x_train_up, y_train_up)
knn_pred = knn_model.predict(x_test)

In [188]:
cm = confusion_matrix(num_y_test, knn_pred)
cm

array([[507,   6,  23,   0,   1,   0],
       [  0, 458,  73,   1,   0,   0],
       [  0, 149, 340,   0,   2,   0],
       [  0,   0,   0, 416,  32,  48],
       [  0,   0,   0,  98, 342,  31],
       [  0,   0,   0,  65,  34, 321]])

#### #2

In [189]:
knn = KNeighborsClassifier(n_neighbors = 3)
knn_model = knn.fit(x_train_up, y_train_up)
knn_pred = knn_model.predict(x_test)

In [190]:
cm = confusion_matrix(num_y_test, knn_pred)
cm

array([[510,   8,  19,   0,   0,   0],
       [  0, 450,  81,   1,   0,   0],
       [  0, 154, 335,   0,   2,   0],
       [  0,   0,   0, 413,  30,  53],
       [  0,   0,   0,  99, 332,  40],
       [  0,   0,   0,  59,  34, 327]])

**Observations:** <br>
The K-Nearest Neighbor models are quite similar to each other. They both do relatively well compared to each other.

### Gaussian Naive Bayes

#### #1

In [191]:
from sklearn.naive_bayes import GaussianNB

gnb = GaussianNB()
gnb_model = gnb.fit(x_train_up, y_train_up)
gnb_pred = gnb_model.predict(x_test)

In [192]:
cm = confusion_matrix(num_y_test, gnb_pred)
cm

array([[343,   0, 191,   0,   3,   0],
       [  7, 444,  66,   0,  15,   0],
       [  5, 103, 376,   0,   7,   0],
       [  0,   0,   0, 418,  35,  43],
       [  0,   0,   0,   9, 450,  12],
       [  0,   0,   0,  79,  83, 258]])

**Observations:** <br>
If Logistic Regression was chosen as one of the classification algorithm for this data, then we would have to use the OneVSRestClassifer or OneVsOneClassifier to handle multiple target classes. But in this case, the algorithms except SVM were capable of handling multiple classes. SVM needed to use the OneVSRestClassifier or OneVSOneClassifer. 

Overall, the SVM model did the best at separating the 6 classes and identifing the correct targets.